rfc/rfc007-api-stability.md

```diff
@@ -0,0 +1,93 @@
+---
+RFC: 007
+Author: Andrew Gerrand <adg@golang.org>
+Status: Draft
+---
+
+# API Stability Policy
+
+## Motivation
+
+The gokit project depends on code maintained by others.
+This includes the Go standard library and sub-repositories
+and other external libraries.
+
+The Go project provides stability guarantees, but the other external libraries
+typically do not. This RFC proposes a standard policy for package authors to
+advertise API stability.
+
+The intention is that the gokit project will require that its dependencies
+adhere to the policy, with the greater goal of improving the Go ecosystem
+as a whole.
+
+## Scope
+
+This policy is for package authors to provide their users with a promise of API
+stability.
+This document is similar to and inspired by the [Go 1 compatibility
+promise](https://golang.org/doc/go1compat).
+
+An author declaring their package "API Stable" makes the following promise:
+
+> We will not change the package's exported API in backward incompatible ways.
+> Future changes to this package will not break dependent code.
+
+### Coverage
+
+The promise of stability includes:
+
+* The package name,
+* Exported type declarations and struct fields (names and types),
+* Exported function and method names, parameters, and return values,
```

Comment:
On 9 March 2015 at 09:50, David Symonds <notifications@github.com> wrote:

> Could you clarify this a bit? If there's an exported function that returns
> an unexported type (as bad a style as that is), is that covered here? To
> what extent? It probably doesn't have to cover the unexported name, but
> does the type identity need to remain stable? (e.g. two exported functions,
> both returning the same unexported type; do they always have to return the
> same type in the future?)
>
> Maybe this should just rule out functions and methods that operate on
> unexported types, whether they be arguments or return values.
>
Just functions and methods? What about unexported names used in type
declarations?

I'm tempted to say "the type identity must remain the same", which seems
like the loosest guarantee of compatibility we can provide. Maybe @robpike
or @griesemer can offer a suggestion.

Also, do you have an opinion on the TODOs about global variables and
constants? Should changes to them break compatibility? I suppose it depends
how opaque the values are.